File: java/logbrew-java/src/main/java/co/logbrew/sdk/HttpTransport.java
package co.logbrew.sdk;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;

/**
 * Dependency-free HTTP transport backed by Java 11's {@link HttpClient}.
 */
public final class HttpTransport implements Transport {
    /**
     * Production LogBrew event intake endpoint used when no endpoint is supplied.
     */
    public static final URI DEFAULT_ENDPOINT = URI.create("https://api.logbrew.co/v1/events");

    private static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(10);
    private static final HttpClient DEFAULT_CLIENT = HttpClient.newBuilder()
        .connectTimeout(DEFAULT_TIMEOUT)
        .build();

    private final URI endpoint;
    private final Map<String, String> headers;
    private final HttpClient client;
    private final Duration requestTimeout;

    /**
     * Creates a transport using the production endpoint and safe default timeout.
     */
    public HttpTransport() {
        this(builder());
    }

    /**
     * Creates a transport for a custom endpoint.
     */
    public HttpTransport(URI endpoint) {
        this(builder().endpoint(endpoint));
    }

    /**
     * Creates a transport for a custom endpoint and extra request headers.
     */
    public HttpTransport(URI endpoint, Map<String, String> headers) {
        this(builder().endpoint(endpoint).headers(headers));
    }

    /**
     * Creates a transport for a custom endpoint, headers, and HTTP client.
     */
    public HttpTransport(URI endpoint, Map<String, String> headers, HttpClient client) {
        this(builder().endpoint(endpoint).headers(headers).client(client));
    }

    private HttpTransport(Builder builder) {
        this.endpoint = validateEndpoint(builder.endpoint);
        this.headers = Collections.unmodifiableMap(copyHeaders(builder.headers));
        this.requestTimeout = validateTimeout(builder.requestTimeout);
        this.client = builder.client == null
            ? defaultClient(this.requestTimeout)
            : builder.client;
    }

    /**
     * Creates a builder for custom endpoint, header, client, and timeout settings.
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Returns the endpoint used for HTTP delivery.
     */
    public URI endpoint() {
        return endpoint;
    }

    /**
     * Returns extra headers copied onto every delivery request.
     */
    public Map<String, String> headers() {
        return headers;
    }

    /**
     * Returns the HTTP client used for delivery.
     */
    public HttpClient client() {
        return client;
    }

    /**
     * Returns the per-request timeout used for delivery.
     */
    public Duration requestTimeout() {
        return requestTimeout;
    }

    @Override
    public TransportResponse send(String apiKey, String body) throws TransportException {
        Validation.requireNonEmpty("api_key", apiKey);
        Objects.requireNonNull(body, "body");

        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder(endpoint)
            .timeout(requestTimeout)
            .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
            .setHeader("content-type", "application/json")
            .setHeader("authorization", "Bearer " + apiKey);

        for (Map.Entry<String, String> header : headers.entrySet()) {
            requestBuilder.setHeader(header.getKey(), header.getValue());
        }

        try {
            HttpResponse<Void> response = client.send(requestBuilder.build(), HttpResponse.BodyHandlers.discarding());
            return new TransportResponse(response.statusCode(), 1);
        } catch (IOException error) {
            throw TransportException.network("http transport failed: " + error.getMessage());
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw TransportException.network("http transport interrupted");
        } catch (IllegalArgumentException error) {
            throw new SdkException("configuration_error", "invalid HTTP transport request: " + error.getMessage());
        }
    }

    private static URI validateEndpoint(URI endpoint) {
        URI resolved = endpoint == null ? DEFAULT_ENDPOINT : endpoint;
        String scheme = resolved.getScheme();
        if (scheme == null) {
            throw new SdkException("configuration_error", "HTTP transport endpoint must include a scheme");
        }
        String normalizedScheme = scheme.toLowerCase(Locale.ROOT);
        if (!"http".equals(normalizedScheme) && !"https".equals(normalizedScheme)) {
            throw new SdkException("configuration_error", "HTTP transport endpoint must use http or https");
        }
        if (resolved.getRawAuthority() == null || resolved.getRawAuthority().trim().isEmpty()) {
            throw new SdkException("configuration_error", "HTTP transport endpoint must include an authority");
        }
        return resolved;
    }

    private static Duration validateTimeout(Duration timeout) {
        Duration resolved = timeout == null ? DEFAULT_TIMEOUT : timeout;
        if (resolved.isZero() || resolved.isNegative()) {
            throw new SdkException("configuration_error", "HTTP transport timeout must be positive");
        }
        return resolved;
    }

    private static HttpClient defaultClient(Duration timeout) {
        if (DEFAULT_TIMEOUT.equals(timeout)) {
            return DEFAULT_CLIENT;
        }
        return HttpClient.newBuilder().connectTimeout(timeout).build();
    }

    private static Map<String, String> copyHeaders(Map<String, String> headers) {
        if (headers == null || headers.isEmpty()) {
            return Collections.emptyMap();
        }
        Map<String, String> copied = new LinkedHashMap<>();
        for (Map.Entry<String, String> header : headers.entrySet()) {
            String name = header.getKey();
            if (name == null) {
                throw new SdkException("configuration_error", "HTTP transport header name must be non-empty");
            }
            if (name.trim().isEmpty()) {
                throw new SdkException("configuration_error", "HTTP transport header name must be non-empty");
            }
            if (header.getValue() == null) {
                throw new SdkException("configuration_error", "HTTP transport header value must be non-null");
            }
            copied.put(name, header.getValue());
        }
        return copied;
    }

    /**
     * Builder for {@link HttpTransport}.
     */
    public static final class Builder {
        private URI endpoint = DEFAULT_ENDPOINT;
        private final Map<String, String> headers = new LinkedHashMap<>();
        private HttpClient client;
        private Duration requestTimeout = DEFAULT_TIMEOUT;

        private Builder() {
        }

        /**
         * Sets the delivery endpoint.
         */
        public Builder endpoint(URI endpoint) {
            this.endpoint = endpoint == null ? DEFAULT_ENDPOINT : endpoint;
            return this;
        }

        /**
         * Adds one extra request header.
         */
        public Builder header(String name, String value) {
            Map<String, String> oneHeader = new LinkedHashMap<>();
            oneHeader.put(name, value);
            this.headers.putAll(copyHeaders(oneHeader));
            return this;
        }

        /**
         * Adds extra request headers.
         */
        public Builder headers(Map<String, String> headers) {
            this.headers.putAll(copyHeaders(headers));
            return this;
        }

        /**
         * Sets the HTTP client used for delivery.
         */
        public Builder client(HttpClient client) {
            this.client = client;
            return this;
        }

        /**
         * Sets the per-request timeout used for delivery.
         */
        public Builder timeout(Duration requestTimeout) {
            this.requestTimeout = requestTimeout;
            return this;
        }

        /**
         * Builds the configured HTTP transport.
         */
        public HttpTransport build() {
            return new HttpTransport(this);
        }
    }
}

File: php/logbrew-php/src/HttpTransport.php
<?php

declare(strict_types=1);

namespace LogBrew;

use Closure;
use Throwable;

/**
 * Dependency-free HTTP transport for sending queued event batches to LogBrew.
 */
final class HttpTransport implements Transport
{
    public const DEFAULT_ENDPOINT = 'https://api.logbrew.co/v1/events';
    public const DEFAULT_TIMEOUT = 10.0;

    /** @var array<string, string> */
    public readonly array $headers;

    /** @var Closure|null */
    private readonly ?Closure $requester;

    /**
     * @param array<string, string> $headers
     */
    public function __construct(
        public readonly string $endpoint = self::DEFAULT_ENDPOINT,
        array $headers = [],
        public readonly float $timeout = self::DEFAULT_TIMEOUT,
        ?callable $requester = null
    ) {
        $this->validateEndpoint($endpoint);
        $this->validateTimeout($timeout);
        $this->headers = $this->copyHeaders($headers);
        $this->requester = $requester === null ? null : Closure::fromCallable($requester);
    }

    /**
     * POST one serialized event batch and return the HTTP status.
     */
    public function send(string $apiKey, string $body): TransportResponse
    {
        LogBrewClient::requireNonEmpty('api_key', $apiKey);
        if ($body === '') {
            throw new SdkError('validation_error', 'body must be non-empty');
        }

        $context = stream_context_create([
            'http' => [
                'method' => 'POST',
                'header' => $this->headerLines($apiKey),
                'content' => $body,
                'ignore_errors' => true,
                'timeout' => $this->timeout,
                'protocol_version' => 1.1,
                'follow_location' => 0,
            ],
        ]);

        try {
            if ($this->requester !== null) {
                $result = ($this->requester)($this->endpoint, $context);
                if ($result instanceof TransportResponse) {
                    return $result;
                }
                if (!is_int($result)) {
                    throw new SdkError('configuration_error', 'HTTP transport requester must return a status code');
                }

                return new TransportResponse($result, 1);
            }

            return new TransportResponse($this->sendWithStreams($context), 1);
        } catch (TransportError|SdkError $error) {
            throw $error;
        } catch (Throwable $error) {
            throw TransportError::network('http transport failed: ' . $error->getMessage());
        }
    }

    /**
     * @return list<string>
     */
    private function headerLines(string $apiKey): array
    {
        $headers = [
            'content-type' => 'application/json',
            'authorization' => 'Bearer ' . $apiKey,
            'connection' => 'close',
        ];

        foreach ($this->headers as $name => $value) {
            $headers[$name] = $value;
        }

        $lines = [];
        foreach ($headers as $name => $value) {
            $lines[] = $name . ': ' . $value;
        }

        return $lines;
    }

    /**
     * @param resource $context
     */
    private function sendWithStreams($context): int
    {
        error_clear_last();
        $responseBody = @file_get_contents($this->endpoint, false, $context);
        $streamError = error_get_last();

        if (function_exists('http_get_last_response_headers')) {
            $headers = http_get_last_response_headers() ?? [];
        } else {
            $definedVariables = get_defined_vars();
            $headers = is_array($definedVariables['http_response_header'] ?? null)
                ? $definedVariables['http_response_header']
                : [];
        }
        $statusCode = $this->statusCodeFromHeaders($headers);
        if ($statusCode !== null) {
            return $statusCode;
        }

        if ($responseBody === false) {
            $message = $streamError === null
                ? 'request failed'
                : $streamError['message'];
            throw TransportError::network('http transport failed: ' . $message);
        }

        throw TransportError::network('http transport failed: missing HTTP status');
    }

    /**
     * @param list<string> $headers
     */
    private function statusCodeFromHeaders(array $headers): ?int
    {
        foreach ($headers as $header) {
            if (preg_match('/^HTTP\/\S+\s+(\d{3})\b/', $header, $matches) === 1) {
                return (int) $matches[1];
            }
        }

        return null;
    }

    private function validateEndpoint(string $endpoint): void
    {
        LogBrewClient::requireNonEmpty('endpoint', $endpoint);
        $parts = parse_url($endpoint);
        if (!is_array($parts)) {
            throw new SdkError('configuration_error', 'invalid HTTP transport endpoint');
        }

        $scheme = strtolower((string) ($parts['scheme'] ?? ''));
        $target = (string) ($parts['host'] ?? '');
        if (($scheme !== 'http' && $scheme !== 'https') || trim($target) === '') {
            throw new SdkError('configuration_error', 'HTTP transport endpoint must use http or https');
        }
    }

    private function validateTimeout(float $timeout): void
    {
        if ($timeout <= 0) {
            throw new SdkError('configuration_error', 'HTTP transport timeout must be positive');
        }
    }

    /**
     * @param array<string, string> $headers
     * @return array<string, string>
     */
    private function copyHeaders(array $headers): array
    {
        $copied = [];
        foreach ($headers as $name => $value) {
            if (trim((string) $name) === '') {
                throw new SdkError('configuration_error', 'HTTP transport header name must be non-empty');
            }
            if (!is_string($value)) {
                throw new SdkError('configuration_error', 'HTTP transport header value must be a string');
            }
            if (str_contains((string) $name, "\r") || str_contains((string) $name, "\n")) {
                throw new SdkError('configuration_error', 'HTTP transport header name must be a single line');
            }
            if (str_contains($value, "\r") || str_contains($value, "\n")) {
                throw new SdkError('configuration_error', 'HTTP transport header value must be a single line');
            }
            $copied[(string) $name] = $value;
        }

        return $copied;
    }
}
